### ISSUE TITLE
Boolean OR, AND, XOR: Reduction operators over integers

### ISSUE THREAD

**Issue by KrishnaPG** [tensorflow/tfjs#7200]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow.js version (you are using): 4.1.0
- Are you willing to contribute it (Yes/No): Would be glad to but do not have the expertise

**Describe the feature and the current behavior/state.**
Kindly add `boolean AND, OR, XOR` as reduction operators over integer typed tensors.

For example, for a 2-dimensional matrix, column-wise boolean AND/OR/XOR reduction should result in a 1-dimensional tensor etc. Similarly row-wise etc.

**Will this change the current api? How?**
New reduction operators may need to be added in addition to the existing `sum`, `prod` etc. operators.

- Three new boolean reduction operators: AND, OR, XOR

**Who will benefit with this feature?**

- Boolean operators enable bit-packing, which will vastly reduce the memory requirement in boolean matrix manipulation senarios. For example, boolean matrix multiplication currently forces us to use "int32" dType, (the bool type is mostly restricted to just logical operators currently it seems).  Enabling boolean reduction operators allows extending the boolean operations further into arithmetic, than just logical.

**Any Other info.**

**Comment by Linchenn** [comment]
Hi @KrishnaPG , do you have use cases for these ops, especially some well-known models? 

Have you found similar APIs from TensorFlow's API? I just found [TensorFlow's bitwise ops](https://www.tensorflow.org/api_docs/python/tf/bitwise/bitwise_and).

**Comment by KrishnaPG** [comment]
Thank you @Linchenn The `tf.bitwise` from the link you have suggested is exactly what is needed - but from reason I could not find the `tf.bitwise` in the JS version (documentation), though. 

Is it available in the JS version internally (but not documented?)

**Comment by Linchenn** [comment]
No, it has not been implemented in TensorFlow.js. It makes sense to implement this.

Do you have use cases for this, or have you seen any popular or well known models that have these ops? It would be appreciated.

**Comment by KrishnaPG** [comment]
I am not aware of ML Models that use these, @Linchenn . I was testing the `tf` for general vector arithmetic over bool and it became very difficult without the boolean operators (Shift, AND, OR, bitCount ...). Since the bool logical operators were already present (ANY, ALL ...), I felt these other operators were missing and hence thought of asking to see if they can be added.

**Comment by Linchenn** [comment]
Got it. Thank you for spotting it out!

**Comment by lcrh** [comment]
This would be great for hashing data, e.g., for bucketing. I'm currently encountering this while trying to work around missing XLA_JIT support for tf.fingerprint.

**Comment by nileshau-afk** [comment]
Hello @KrishnaPG  ,

Thank you for bringing this to our attention and for providing a comprehensive overview. We appreciate the time you invested in creating this issue.

Currently, our internal development team is concentrated on other scheduled milestones, so we cannot prioritize this task immediately. However, we would love to see this move forward as a community-led effort.

If you’d like to submit a fix, please fork the repo and open a PR following the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute), [Tfjs Contributions](https://github.com/tensorflow/tfjs/blob/master/tfjs-vis/CONTRIBUTING.md#how-to-contribute) and be sure to link this issue in your PR to avoid duplicate work, kindly leave a comment here if you plan to take this on.

Thank you again for your interest and time.

**Comment by nileshau-afk** [comment]
Hello @KrishnaPG  ,

Thank you for opening this issue. Since this issue has been open for a long time, we are closing this issue for now.
If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.

Thank you for your understanding.
